File: io.openems.edge.core/src/io/openems/edge/core/predictormanager/Get24HoursPredictionRequest.java
package io.openems.edge.core.predictormanager;

import java.util.ArrayList;
import java.util.List;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
import io.openems.common.jsonrpc.base.JsonrpcRequest;
import io.openems.common.types.ChannelAddress;
import io.openems.common.utils.JsonUtils;

/**
 * Wraps a JSON-RPC Request to query a 24 Hours Prediction.
 *
 * <pre>
 * {
 *   "jsonrpc": "2.0",
 *   "id": "UUID",
 *   "method": "get24HoursPrediction",
 *   "params": {
 *   	"channels": string[]
 *   }
 * }
 * </pre>
 */
public class Get24HoursPredictionRequest extends JsonrpcRequest {

	public static final String METHOD = "get24HoursPrediction";

	/**
	 * Create {@link Get24HoursPredictionRequest} from a template
	 * {@link JsonrpcRequest}.
	 *
	 * @param r the template {@link JsonrpcRequest}
	 * @return the {@link Get24HoursPredictionRequest}
	 * @throws OpenemsNamedException on parse error
	 */
	public static Get24HoursPredictionRequest from(JsonrpcRequest r) throws OpenemsNamedException {
		var p = r.getParams();
		var cs = JsonUtils.getAsJsonArray(p, "channels");
		List<ChannelAddress> channels = new ArrayList<>();
		for (JsonElement c : cs) {
			channels.add(ChannelAddress.fromString(JsonUtils.getAsString(c)));
		}
		return new Get24HoursPredictionRequest(r, channels);
	}

	private final List<ChannelAddress> channels;

	public Get24HoursPredictionRequest(List<ChannelAddress> channels) {
		super(METHOD);
		this.channels = channels;
	}

	private Get24HoursPredictionRequest(JsonrpcRequest request, List<ChannelAddress> channels) {
		super(request, METHOD);
		this.channels = channels;
	}

	@Override
	public JsonObject getParams() {
		var channels = new JsonArray();
		for (ChannelAddress channel : this.channels) {
			channels.add(channel.toString());
		}
		return JsonUtils.buildJsonObject() //
				.add("channels", channels) //
				.build();
	}

	/**
	 * Gets the {@link ChannelAddress}es.
	 *
	 * @return a list of {@link ChannelAddress}
	 */
	public List<ChannelAddress> getChannels() {
		return this.channels;
	}

}

File: ui/src/app/shared/jsonrpc/request/get24HoursPredictionRequest.ts
import { ChannelAddress } from "../../../shared/type/channeladdress";
import { format } from 'date-fns';
import { JsonrpcRequest } from "../base";
import { JsonRpcUtils } from "../jsonrpcutils";

/**
 * Represents a JSON-RPC Request to query a 24 Hours Prediction.
 * 
 * <pre>
 * {
 *   "jsonrpc": "2.0",
 *   "id": UUID,
 *   "method": "get24HoursPredictionRequest",
 *   "params": {
 *     "channels": ChannelAddress[]
 *   }
 * }
 * </pre>
 */
export class Get24HoursPredictionRequest extends JsonrpcRequest {

    private static METHOD: string = "get24HoursPrediction";

    public constructor(
        private channels: ChannelAddress[]
    ) {
        super(Get24HoursPredictionRequest.METHOD, {
            channels: JsonRpcUtils.channelsToStringArray(channels)
        });
        // delete local fields, otherwise they are sent with the JSON-RPC Request
        delete this.channels;
    }

}